File: clients/apple/MultiLiveTV/Services/MacCMSModels.swift
import Foundation

struct VodTypeRaw: Decodable {
    let typeId: Int
    let typeName: String

    enum CodingKeys: String, CodingKey {
        case typeId = "type_id"
        case typeName = "type_name"
    }
}

struct VodItemRaw {
    let vodId: String
    let vodName: String
    let vodPic: String
    let vodRemarks: String
    let vodYear: String
    let vodArea: String
    let vodClass: String
    let vodBlurb: String
    let vodContent: String
    let vodPlayFrom: String
    let vodPlayURL: String
    let typeId: Int
    let typeName: String
    let vodTime: Int

    func toVodItem(variants: [VodVariant]? = nil, primarySourceId: Int? = nil) -> VodItem {
        VodItem(
            vodId: vodId,
            vodName: vodName,
            vodPic: vodPic,
            vodRemarks: vodRemarks.isEmpty ? nil : vodRemarks,
            vodBlurb: vodBlurb.isEmpty ? nil : vodBlurb,
            vodContent: vodContent.isEmpty ? nil : vodContent,
            typeName: typeName.isEmpty ? nil : typeName,
            vodClass: vodClass.isEmpty ? nil : vodClass,
            variants: variants,
            primarySourceId: primarySourceId,
            vodTime: vodTime
        )
    }
}

struct MacCMSListResponse {
    let code: Int
    let msg: String
    let page: Int
    let pageCount: Int
    let limit: String
    let total: Int
    let list: [VodItemRaw]
    let types: [VodTypeRaw]
}

struct MacCMSDetailResponse {
    let code: Int
    let msg: String
    let list: [VodItemRaw]
}

struct MergeableVodItem {
    let item: VodItemRaw
    let sourceId: Int
    let sourceName: String
}

enum MacCMSJSONParser {
    static func parseList(_ data: Data) throws -> MacCMSListResponse {
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw DecodingError.dataCorrupted(.init(codingPath: [], debugDescription: "invalid json"))
        }
        let list = parseVodItems(json["list"])
        let types = parseTypes(json["class"])
        return MacCMSListResponse(
            code: flexInt(json["code"]) ?? 0,
            msg: json["msg"] as? String ?? "",
            page: flexInt(json["page"]) ?? 1,
            pageCount: flexInt(json["pagecount"]) ?? 1,
            limit: flexString(json["limit"]) ?? "",
            total: flexInt(json["total"]) ?? 0,
            list: list,
            types: types
        )
    }

    static func parseDetail(_ data: Data) throws -> MacCMSDetailResponse {
        guard let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            throw DecodingError.dataCorrupted(.init(codingPath: [], debugDescription: "invalid json"))
        }
        return MacCMSDetailResponse(
            code: flexInt(json["code"]) ?? 0,
            msg: json["msg"] as? String ?? "",
            list: parseVodItems(json["list"])
        )
    }

    private static func parseVodItems(_ value: Any?) -> [VodItemRaw] {
        guard let array = value as? [[String: Any]] else { return [] }
        return array.compactMap(parseVodItem)
    }

    private static func parseTypes(_ value: Any?) -> [VodTypeRaw] {
        guard let array = value as? [[String: Any]] else { return [] }
        return array.compactMap { dict in
            guard let typeId = flexInt(dict["type_id"]),
                  let typeName = dict["type_name"] as? String else { return nil }
            return VodTypeRaw(typeId: typeId, typeName: typeName)
        }
    }

    private static func parseVodItem(_ dict: [String: Any]) -> VodItemRaw? {
        guard let vodName = dict["vod_name"] as? String else { return nil }
        let vodId = flexString(dict["vod_id"]) ?? ""
        let vodTime = parseVodTime(dict["vod_time"])
        let vodTimeAdd = parseVodTime(dict["vod_time_add"])
        return VodItemRaw(
            vodId: vodId,
            vodName: vodName,
            vodPic: dict["vod_pic"] as? String ?? "",
            vodRemarks: dict["vod_remarks"] as? String ?? "",
            vodYear: dict["vod_year"] as? String ?? "",
            vodArea: dict["vod_area"] as? String ?? "",
            vodClass: dict["vod_class"] as? String ?? "",
            vodBlurb: dict["vod_blurb"] as? String ?? "",
            vodContent: dict["vod_content"] as? String ?? "",
            vodPlayFrom: dict["vod_play_from"] as? String ?? "",
            vodPlayURL: dict["vod_play_url"] as? String ?? "",
            typeId: flexInt(dict["type_id"]) ?? 0,
            typeName: dict["type_name"] as? String ?? "",
            vodTime: vodTime > 0 ? vodTime : vodTimeAdd
        )
    }

    static func parseVodTime(_ value: Any?) -> Int {
        switch value {
        case let n as Int:
            return normalizeUnix(n)
        case let n as Double:
            return normalizeUnix(Int(n))
        case let n as NSNumber:
            return normalizeUnix(n.intValue)
        case let s as String:
            let trimmed = s.trimmingCharacters(in: .whitespacesAndNewlines)
            if trimmed.isEmpty { return 0 }
            if let n = Int(trimmed) { return normalizeUnix(n) }
            return parseDateTimeString(trimmed)
        default:
            return 0
        }
    }

    private static func normalizeUnix(_ value: Int) -> Int {
        if value <= 0 { return 0 }
        if value > 10_000_000_000 { return value / 1000 }
        return value
    }

    private static func parseDateTimeString(_ raw: String) -> Int {
        let parts = raw.split { !$0.isNumber }.compactMap { Int($0) }
        guard parts.count >= 3 else { return 0 }

        var comps = DateComponents()
        comps.calendar = Calendar(identifier: .gregorian)
        comps.timeZone = TimeZone(identifier: "Asia/Shanghai")
        comps.year = parts[0]
        comps.month = parts[1]
        comps.day = parts[2]
        comps.hour = parts.count > 3 ? parts[3] : 0
        comps.minute = parts.count > 4 ? parts[4] : 0
        comps.second = parts.count > 5 ? parts[5] : 0

        guard let date = comps.date else { return 0 }
        return Int(date.timeIntervalSince1970)
    }

    private static func flexString(_ value: Any?) -> String? {
        switch value {
        case let s as String: return s
        case let n as Int: return String(n)
        case let n as Double: return String(Int(n))
        case let n as NSNumber: return n.stringValue
        default: return nil
        }
    }

    private static func flexInt(_ value: Any?) -> Int? {
        switch value {
        case let n as Int: return n
        case let n as Double: return Int(n)
        case let s as String: return Int(s)
        case let n as NSNumber: return n.intValue
        default: return nil
        }
    }
}

File: clients/android/core/src/main/kotlin/com/heibaimiao/multilivetv/maccms/MacCMSModels.kt
package com.heibaimiao.multilivetv.maccms

import com.heibaimiao.multilivetv.model.VodItem
import com.heibaimiao.multilivetv.model.VodVariant
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.time.LocalDateTime
import java.time.ZoneId

data class VodTypeRaw(
    val typeId: Int,
    val typeName: String,
)

data class VodItemRaw(
    val vodId: String,
    val vodName: String,
    val vodPic: String,
    val vodRemarks: String,
    val vodYear: String,
    val vodArea: String,
    val vodClass: String,
    val vodBlurb: String,
    val vodContent: String,
    val vodActor: String = "",
    val vodDirector: String = "",
    val vodPlayFrom: String,
    val vodPlayURL: String,
    val typeId: Int,
    val typeName: String,
    val vodTime: Int,
) {
    fun toVodItem(variants: List<VodVariant>? = null, primarySourceId: Int? = null): VodItem = VodItem(
        vodId = vodId,
        vodName = vodName,
        vodPic = vodPic,
        vodRemarks = vodRemarks.ifEmpty { null },
        vodBlurb = vodBlurb.ifEmpty { null },
        vodContent = vodContent.ifEmpty { null },
        typeName = typeName.ifEmpty { null },
        vodClass = vodClass.ifEmpty { null },
        vodYear = vodYear.ifEmpty { null },
        variants = variants,
        primarySourceId = primarySourceId,
        vodTime = vodTime,
    )

    fun withVodTime(time: Int): VodItemRaw = copy(vodTime = time)
}

data class MacCMSListResponse(
    val code: Int,
    val msg: String,
    val page: Int,
    val pageCount: Int,
    val limit: String,
    val total: Int,
    val list: List<VodItemRaw>,
    val types: List<VodTypeRaw>,
)

data class MacCMSDetailResponse(
    val code: Int,
    val msg: String,
    val list: List<VodItemRaw>,
)

data class MergeableVodItem(
    val item: VodItemRaw,
    val sourceId: Int,
    val sourceName: String,
)

object MacCMSJsonParser {
    fun parseList(root: JsonObject): MacCMSListResponse = MacCMSListResponse(
        code = flexInt(root["code"]) ?: 0,
        msg = root["msg"]?.jsonPrimitive?.contentOrNull.orEmpty(),
        page = flexInt(root["page"]) ?: 1,
        pageCount = flexInt(root["pagecount"]) ?: 1,
        limit = flexString(root["limit"]).orEmpty(),
        total = flexInt(root["total"]) ?: 0,
        list = parseVodItems(root["list"]),
        types = parseTypes(root["class"]),
    )

    fun parseDetail(root: JsonObject): MacCMSDetailResponse = MacCMSDetailResponse(
        code = flexInt(root["code"]) ?: 0,
        msg = root["msg"]?.jsonPrimitive?.contentOrNull.orEmpty(),
        list = parseVodItems(root["list"]),
    )

    fun parseVodTime(value: JsonElement?): Int = when (value) {
        null -> 0
        is JsonPrimitive -> {
            val intValue = flexInt(value)
            if (intValue != null) normalizeUnix(intValue) else parseDateTimeString(value.content.trim())
        }
        else -> 0
    }

    private fun parseVodItems(value: JsonElement?): List<VodItemRaw> {
        val array = value as? JsonArray ?: return emptyList()
        return array.mapNotNull { element ->
            val obj = element as? JsonObject ?: return@mapNotNull null
            parseVodItem(obj)
        }
    }

    private fun parseTypes(value: JsonElement?): List<VodTypeRaw> {
        val array = value as? JsonArray ?: return emptyList()
        return array.mapNotNull { element ->
            val obj = element as? JsonObject ?: return@mapNotNull null
            val typeId = flexInt(obj["type_id"]) ?: return@mapNotNull null
            val typeName = obj["type_name"]?.jsonPrimitive?.contentOrNull ?: return@mapNotNull null
            VodTypeRaw(typeId, typeName)
        }
    }

    private fun parseVodItem(dict: JsonObject): VodItemRaw? {
        val vodName = dict["vod_name"]?.jsonPrimitive?.contentOrNull ?: return null
        val vodTime = parseVodTime(dict["vod_time"])
        val vodTimeAdd = parseVodTime(dict["vod_time_add"])
        return VodItemRaw(
            vodId = flexString(dict["vod_id"]).orEmpty(),
            vodName = vodName,
            vodPic = dict["vod_pic"]?.jsonPrimitive?.contentOrNull.orEmpty(),
            vodRemarks = dict["vod_remarks"]?.jsonPrimitive?.contentOrNull.orEmpty(),
            vodYear = dict["vod_year"]?.jsonPrimitive?.contentOrNull.orEmpty(),
            vodArea = dict["vod_area"]?.jsonPrimitive?.contentOrNull.orEmpty(),
            vodClass = dict["vod_class"]?.jsonPrimitive?.contentOrNull.orEmpty(),
            vodBlurb = dict["vod_blurb"]?.jsonPrimitive?.contentOrNull.orEmpty(),
            vodContent = dict["vod_content"]?.jsonPrimitive?.contentOrNull.orEmpty(),
            vodActor = dict["vod_actor"]?.jsonPrimitive?.contentOrNull.orEmpty(),
            vodDirector = dict["vod_director"]?.jsonPrimitive?.contentOrNull.orEmpty(),
            vodPlayFrom = dict["vod_play_from"]?.jsonPrimitive?.contentOrNull.orEmpty(),
            vodPlayURL = dict["vod_play_url"]?.jsonPrimitive?.contentOrNull.orEmpty(),
            typeId = flexInt(dict["type_id"]) ?: 0,
            typeName = dict["type_name"]?.jsonPrimitive?.contentOrNull.orEmpty(),
            vodTime = if (vodTime > 0) vodTime else vodTimeAdd,
        )
    }

    private fun normalizeUnix(value: Int): Int = when {
        value <= 0 -> 0
        value > 10_000_000_000 -> value / 1000
        else -> value
    }

    private fun parseDateTimeString(raw: String): Int {
        val parts = Regex("""\d+""").findAll(raw).map { it.value.toInt() }.toList()
        if (parts.size < 3) return 0
        return try {
            val date = LocalDateTime.of(
                parts[0],
                parts[1],
                parts[2],
                parts.getOrElse(3) { 0 },
                parts.getOrElse(4) { 0 },
                parts.getOrElse(5) { 0 },
            )
            date.atZone(ZoneId.of("Asia/Shanghai")).toEpochSecond().toInt()
        } catch (_: Exception) {
            0
        }
    }

    fun flexString(value: JsonElement?): String? = when (value) {
        null -> null
        is JsonPrimitive -> if (value.isString) value.content else value.content
        else -> null
    }

    fun flexInt(value: JsonElement?): Int? = when (value) {
        null -> null
        is JsonPrimitive -> value.content.toDoubleOrNull()?.toInt()
        else -> null
    }
}
